core/src/main/java/hudson/util/AtomicFileWriter.java

```diff
@@ -108,35 +119,72 @@ public void close() throws IOException {
      * the {@link #commit()} is called, to simplify coding.
      */
     public void abort() throws IOException {
-        close();
-        tmpFile.delete();
+        closeAndDeleteTempFile();
     }
 
     public void commit() throws IOException {
         close();
-        if (destFile.exists()) {
+        try {
+            // Try to make an atomic move.
+            Files.move(tmpPath, destPath, StandardCopyOption.ATOMIC_MOVE);
+        } catch (IOException e) {
+            // If it falls here that can mean many things. Either that the atomic move is not supported,
+            // or something wrong happened. Anyway, let's try to be over-diagnosing
+            LOGGER.log(Level.WARNING, "Unable to move atomically, falling back to non-atomic move.");
+            if (destPath.toFile().exists()) {
+                LOGGER.log(Level.WARNING, "The target file {0} was already existing?!?", destPath);
+            }
+
             try {
-                Util.deleteFile(destFile);
-            } catch (IOException x) {
-                tmpFile.delete();
-                throw x;
+                Files.move(tmpPath, destPath, StandardCopyOption.REPLACE_EXISTING);
+            } catch (IOException e1) {
+                e1.addSuppressed(e);
+                LOGGER.log(Level.SEVERE, "Unable to move {0} to {1}. Attempting to delete {1} and abandoning.",
```

Comment:
:bug:  ` Attempting to delete {1}` should be ` Attempting to delete {0}`

Reply:
Fixed